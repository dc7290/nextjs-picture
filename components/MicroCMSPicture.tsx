import Image, { ImageProps } from 'next/image'
import { ImageConfigContext } from 'next/dist/shared/lib/image-config-context'
import { useContext } from 'react'

type ArtDirective = {
  media: string
}

type Props = {
  artDirevtives?: ArtDirective[]
} & Omit<ImageProps, 'blurDataURL' | 'loader'>

const MicroCMSPicture = ({ src, ...props }: Props) => {
  const nextImageConfig = useContext(ImageConfigContext)

  return (
    <picture>
      <Image src={src} {...props} />
    </picture>
  )
}

export default MicroCMSPicture
