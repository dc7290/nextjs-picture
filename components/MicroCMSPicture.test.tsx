import '@testing-library/jest-dom'

import { render, screen } from '@testing-library/react'
import { NextConfig } from 'next'
import { imageConfigDefault } from 'next/dist/shared/lib/image-config'
import { ReactElement } from 'react'

import { getSources, Sources } from './MicroCMSPicture'

jest.mock('next/head', () => {
  return {
    __esModule: true,
    default: ({ children }: { children: Array<ReactElement> }) => {
      return <>{children}</>
    },
  }
})

const createConfig = (customConfig?: NextConfig['images']) => {
  return Object.assign(imageConfigDefault, customConfig)
}

const defaultSources = getSources({
  config: createConfig({
    deviceSizes: [640, 1200],
  }),
  src: 'https://remote-image/image.png',
  preloadFormat: 'image/webp',
})
const artDirevtionSources = getSources({
  config: createConfig({
    deviceSizes: [640, 1200],
  }),
  artDirevtives: [
    {
      src: 'https://remote-image/image_lg.png',
      media: '(min-width: 1024px)',
      width: 1920,
      height: 1200,
    },
    {
      src: 'https://remote-image/image_md.png',
      media: '(min-width: 768px)',
      width: 1920,
      height: 800,
    },
  ],
  src: 'https://remote-image/image.png',
  preloadFormat: 'image/webp',
})

describe('MicroCMSPicture', () => {
  describe('getSources', () => {
    it('デフォルトの動作', () => {
      expect(defaultSources.sources).toStrictEqual([
        {
          height: undefined,
          srcSet:
            'https://remote-image/image.png?fit=max&w=640&q=75&fm=avif 640w, https://remote-image/image.png?fit=max&w=1200&q=75&fm=avif 1200w',
          type: 'image/avif',
          width: undefined,
        },
        {
          height: undefined,
          srcSet:
            'https://remote-image/image.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image.png?fit=max&w=1200&q=75&fm=webp 1200w',
          type: 'image/webp',
          width: undefined,
        },
      ])
      expect(defaultSources.preloadLinks).toStrictEqual([
        {
          srcSet:
            'https://remote-image/image.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image.png?fit=max&w=1200&q=75&fm=webp 1200w',
          type: 'image/webp',
        },
      ])
    })
    it('アートディレクションの動作', () => {
      expect(artDirevtionSources.sources).toStrictEqual([
        {
          srcSet:
            'https://remote-image/image_lg.png?fit=max&w=640&q=75&fm=avif 640w, https://remote-image/image_lg.png?fit=max&w=1200&q=75&fm=avif 1200w',
          type: 'image/avif',
          media: '(min-width: 1024px)',
          width: 1920,
          height: 1200,
        },
        {
          srcSet:
            'https://remote-image/image_lg.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image_lg.png?fit=max&w=1200&q=75&fm=webp 1200w',
          type: 'image/webp',
          media: '(min-width: 1024px)',
          width: 1920,
          height: 1200,
        },
        {
          srcSet:
            'https://remote-image/image_lg.png?fit=max&w=640&q=75 640w, https://remote-image/image_lg.png?fit=max&w=1200&q=75 1200w',
          media: '(min-width: 1024px)',
          width: 1920,
          height: 1200,
        },
        {
          srcSet:
            'https://remote-image/image_md.png?fit=max&w=640&q=75&fm=avif 640w, https://remote-image/image_md.png?fit=max&w=1200&q=75&fm=avif 1200w',
          type: 'image/avif',
          media: '(min-width: 768px)',
          width: 1920,
          height: 800,
        },
        {
          srcSet:
            'https://remote-image/image_md.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image_md.png?fit=max&w=1200&q=75&fm=webp 1200w',
          type: 'image/webp',
          media: '(min-width: 768px)',
          width: 1920,
          height: 800,
        },
        {
          srcSet:
            'https://remote-image/image_md.png?fit=max&w=640&q=75 640w, https://remote-image/image_md.png?fit=max&w=1200&q=75 1200w',
          media: '(min-width: 768px)',
          width: 1920,
          height: 800,
        },
        {
          height: undefined,
          srcSet:
            'https://remote-image/image.png?fit=max&w=640&q=75&fm=avif 640w, https://remote-image/image.png?fit=max&w=1200&q=75&fm=avif 1200w',
          type: 'image/avif',
          width: undefined,
        },
        {
          height: undefined,
          srcSet:
            'https://remote-image/image.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image.png?fit=max&w=1200&q=75&fm=webp 1200w',
          type: 'image/webp',
          width: undefined,
        },
      ])
      expect(artDirevtionSources.preloadLinks).toStrictEqual([
        {
          media: '(min-width: 1024px)',
          srcSet:
            'https://remote-image/image_lg.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image_lg.png?fit=max&w=1200&q=75&fm=webp 1200w',
          type: 'image/webp',
        },
        {
          media: '(min-width: 768px)',
          srcSet:
            'https://remote-image/image_md.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image_md.png?fit=max&w=1200&q=75&fm=webp 1200w',
          type: 'image/webp',
        },
      ])
    })
  })

  it('デフォルトの動作', () => {
    render(<Sources {...defaultSources} />)
    expect(screen.getAllByTestId('source')).toMatchSnapshot()
  })
  it('アートディレクションの動作', () => {
    render(<Sources {...artDirevtionSources} />)
    expect(screen.getAllByTestId('source')).toMatchSnapshot()
  })

  it('デフォルトの動作: priorityを設定する', () => {
    render(<Sources {...defaultSources} priority />)
    const preloadLink = defaultSources.preloadLinks[0]
    expect(screen.getByTestId('link')).toHaveAttribute(
      'imagesrcset',
      preloadLink.srcSet
    )
    expect(screen.getByTestId('link')).toHaveAttribute('type', preloadLink.type)
  })
  it('アートディレクションの動作: priorityを設定する', () => {
    render(<Sources {...artDirevtionSources} priority />)
    screen.getAllByTestId('link').forEach((link, i) => {
      const preloadLink = artDirevtionSources.preloadLinks[i]
      expect(link).toHaveAttribute('imagesrcset', preloadLink.srcSet)
      expect(link).toHaveAttribute('type', preloadLink.type)
    })
  })
})
