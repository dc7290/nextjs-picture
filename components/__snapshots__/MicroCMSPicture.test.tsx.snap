// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MicroCMSPicture アートディレクションの動作 1`] = `
Array [
  <source
    data-testid="source"
    height="1200"
    media="(min-width: 1024px)"
    srcset="https://remote-image/image_lg.png?fit=max&w=640&q=75&fm=avif 640w, https://remote-image/image_lg.png?fit=max&w=1200&q=75&fm=avif 1200w"
    type="image/avif"
    width="1920"
  />,
  <source
    data-testid="source"
    height="1200"
    media="(min-width: 1024px)"
    srcset="https://remote-image/image_lg.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image_lg.png?fit=max&w=1200&q=75&fm=webp 1200w"
    type="image/webp"
    width="1920"
  />,
  <source
    data-testid="source"
    height="1200"
    media="(min-width: 1024px)"
    srcset="https://remote-image/image_lg.png?fit=max&w=640&q=75 640w, https://remote-image/image_lg.png?fit=max&w=1200&q=75 1200w"
    width="1920"
  />,
  <source
    data-testid="source"
    height="800"
    media="(min-width: 768px)"
    srcset="https://remote-image/image_md.png?fit=max&w=640&q=75&fm=avif 640w, https://remote-image/image_md.png?fit=max&w=1200&q=75&fm=avif 1200w"
    type="image/avif"
    width="1920"
  />,
  <source
    data-testid="source"
    height="800"
    media="(min-width: 768px)"
    srcset="https://remote-image/image_md.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image_md.png?fit=max&w=1200&q=75&fm=webp 1200w"
    type="image/webp"
    width="1920"
  />,
  <source
    data-testid="source"
    height="800"
    media="(min-width: 768px)"
    srcset="https://remote-image/image_md.png?fit=max&w=640&q=75 640w, https://remote-image/image_md.png?fit=max&w=1200&q=75 1200w"
    width="1920"
  />,
  <source
    data-testid="source"
    srcset="https://remote-image/image.png?fit=max&w=640&q=75&fm=avif 640w, https://remote-image/image.png?fit=max&w=1200&q=75&fm=avif 1200w"
    type="image/avif"
  />,
  <source
    data-testid="source"
    srcset="https://remote-image/image.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image.png?fit=max&w=1200&q=75&fm=webp 1200w"
    type="image/webp"
  />,
]
`;

exports[`MicroCMSPicture デフォルトの動作 1`] = `
Array [
  <source
    data-testid="source"
    srcset="https://remote-image/image.png?fit=max&w=640&q=75&fm=avif 640w, https://remote-image/image.png?fit=max&w=1200&q=75&fm=avif 1200w"
    type="image/avif"
  />,
  <source
    data-testid="source"
    srcset="https://remote-image/image.png?fit=max&w=640&q=75&fm=webp 640w, https://remote-image/image.png?fit=max&w=1200&q=75&fm=webp 1200w"
    type="image/webp"
  />,
]
`;
